trieste/acquisition/optimizer.py

```diff
@@ -140,3 +143,44 @@ def target_func_with_vectorized_inputs(
         return tf.reshape(vectorized_points, [batch_size, -1])  # [B, D]
 
     return optimizer
+
+
+def batchify_greedy(
+    batch_size_one_optimizer: AcquisitionOptimizer[SP],
+    batch_size: int,
+) -> AcquisitionOptimizer[SP]:
+    """
+    A wrapper around our :const:`AcquisitionOptimizer`s. This class wraps a
+    :const:`AcquisitionOptimizer` to allow it to approximately optimize batch acquisition functions
+        by greedily choosing each batch point in sequence, i.e choosing the point that provides
+        the largest increase in the acquisiton function over the current (partial) batch.
+    :param batch_size_one_optimizer: An optimizer that returns only batch size one, i.e. produces a
+        single point with shape [1, D].
+    :param batch_size: The number of points in the batch.
+    :return: An :const:`AcquisitionOptimizer` that will provide a batch of points with shape [B, D].
+    """
+    tf.debugging.assert_positive(batch_size)
+
+    def optimizer(search_space: SP, f: AcquisitionFunction) -> TensorType:
+
+        current_batch = batch_size_one_optimizer(search_space, f)  # [1, D]
+
+        for i in range(1, batch_size):
+
+            def target_func_with_fixed_partial_batch(
+                x: TensorType,
+            ) -> TensorType:  # [..., 1, D] -> [..., 1]
+
+                current_batch_tiled = tf.repeat(current_batch[None, :, :], x.shape[-3], axis=-3)
+                x_with_current_batch = tf.concat([current_batch_tiled, x], -2)
+                return f(x_with_current_batch)
+
+            next_batch_point = batch_size_one_optimizer(
+                search_space, target_func_with_fixed_partial_batch
+            )  # [1, D]
+
+            current_batch = tf.concat([current_batch, next_batch_point], axis=0)  # [i+1, D]
```

Comment:
you might want to be careful here with mutable variables: will this work if someone wraps it with `@tf.function`?